### ISSUE TITLE
How to test using flags?

### ISSUE THREAD

**Issue by wallacepf** [KEINOS/Hello-Cobra#6]
First of all, I have following my docs/blogs/posts trying to find a way to test a simple CLI application and yours is the best so, thank you very much. My question is, how can I implement flags in the way you're doing the tests? 

I have tried to do the following once I have a flag receiving a string to disable some fmt.Scan() I have in my code to confirm my actions. Seems like my flag gets ignored.

```
func TestCreateWallet(t *testing.T) {
	var expect = "ok"
	var actual = capturer.CaptureStdout(func() {
		var cmdTmp = &cobra.Command{}
		var argsTmp = []string{"arg1", "--testing", "enabled"}
		createWalletCmd.PreRun(cmdTmp, argsTmp)
	})

	assert.Equal(t, expect, actual, "Result as expected")
}
```

Any help will be highly appreciated and by the way, sry for the noob question. This is my first week coding in go.

**Comment by KEINOS** [comment]
Wasn't expecting that it would be seen by someone, thank you instead!!! 

> how can I implement flags

Good point!

I'll try what can I do to implement tests for option flags!🤞

To make it clear, all we need here is to test:
- if the "`--testing`" boolean flag is set then the code notice it, right?

**Comment by wallacepf** [comment]
> Wasn't expecting that it would be seen by someone, thank you instead!!!
> 
> > how can I implement flags
> 
> Good point!
> 
> I'll try what can I do to implement tests for option flags!🤞
> 
> To make it clear, all we need here is to test:
> 
> * if the "`--testing`" boolean flag is set then the code notice it, right?

Flag + Value, like 

`--myflag myvalue`

> Wasn't expecting that it would be seen by someone, thank you instead!!!

Once the information is available in the internet, we shall consume it!

Even not having experience in Go, I'll try my best to find a way to do that. if I found the solution, I'll do a PR.

**Comment by KEINOS** [comment]
@wallacepf 

I think I found one way of; **how to provide command options(args and flags) in the tests**.

## TL; DR

In the test function, copy/clone the `rootCmd` and override the variable by using `SetArgs()`, then `Execute` the cloned command. 

## TS; DR

```go
func TestCreateWallet(t *testing.T) {
	// copy the entire rootCmd
	var cmdTmp = *rootCmd

	// Replace the entire argument
	var argsTmp = []string{"prerun", "arg1", "--testing", "enabled"}
	cmdTmp.SetArgs(argsTmp)

	// Run test
	var expect = "OK\n"
	var actual = capturer.CaptureStdout(func() {
		cmdTmp.Execute()
	})
	assert.Equal(t, expect, actual, "The two words should be the same.")
}
```

> `SetArgs` sets arguments for the command. It is set to `os.Args[1:]` by default, if desired, can be overridden particularly useful when testing.
>
> - https://github.com/spf13/cobra/blob/7df62f7668c7fbeafd6dbb45ad0294f4ea08c0ec/command.go#L227-L231

I did add a [sample of using flags](https://github.com/KEINOS/Hello-Cobra/blob/b74a143cbe092d241c46a85b5e93ff9dc11d189f/cmd/flag.go) and [its tests](https://github.com/KEINOS/Hello-Cobra/blob/b74a143cbe092d241c46a85b5e93ff9dc11d189f/cmd/flag_test.go) in https://github.com/KEINOS/Hello-Cobra/pull/9, but we need to add this one too.

**Comment by KEINOS** [comment]
I found this sample repo very useful.

- https://github.com/tkit/go-cmd-example @ GitHub

It defines the sub-command as a function rather than declaring it into a variable.

```go
package cmd

import (
	"github.com/spf13/cobra"
)

func NewCmdShow() *cobra.Command {
	type Options struct {
		Optint int    `validate:"min=0,max=10"`
		Optstr string `validate:"required,alphanum"`
	}

	var (
		o = &Options{}
	)

	cmd := &cobra.Command{
		Use:   "show",
		Short: "A brief description of your command",
		PreRunE: func(cmd *cobra.Command, args []string) error {
			return validateParams(*o)
		},
		Run: func(cmd *cobra.Command, args []string) {
			cmd.Printf("show called: optint: %d, optstr: %s", o.Optint, o.Optstr)
		},
		SilenceErrors: true,
		SilenceUsage:  true,
	}
	cmd.Flags().IntVarP(&o.Optint, "int", "i", 0, "int option")
	cmd.Flags().StringVarP(&o.Optstr, "str", "s", "", "string option")

	return cmd
}

func init() {
}
```

This way seems to ease testing.

```go
package cmd

import (
	"bytes"
	"github.com/mattn/go-shellwords"
	"testing"
)

func TestShow(t *testing.T) {
	cases := []struct {
		command string
		want    string
		errWant string
	}{
		{command: "cmd-test show", want: "", errWant: "Parameter error: Optstr is required"},
		{command: "cmd-test show --int 10", want: "", errWant: "Parameter error: Optstr is required"},
		{command: "cmd-test show --str test", want: "show called: optint: 0, optstr: test"},
		{command: "cmd-test show --str \"test1 test2\"", want: "", errWant: "Parameter error: Optstr is not valid test1 test2"},
		{command: "cmd-test show --int 1000 --str 123", want: "", errWant: "Parameter error: Optint cannot be greater than 10"},
		{command: "cmd-test show --int -1 --str abc", want: "", errWant: "Parameter error: Optint must be greater than 0"},
		{command: "cmd-test show --int 1 --str abc", want: "show called: optint: 1, optstr: abc"},
	}

	for _, c := range cases {
		buf := new(bytes.Buffer)
		cmd := NewCmdRoot()
		cmd.SetOutput(buf)
		cmdArgs, err := shellwords.Parse(c.command)
		if err != nil {
			t.Fatalf("args parse error: %+v\n", err)
		}
		cmd.SetArgs(cmdArgs[1:])
		if err := cmd.Execute(); err != nil {
			if c.errWant != err.Error() {
				t.Errorf("unexpected error response: errWant:%+v, get:%+v", c.errWant, err.Error())
			}
		}

		get := buf.String()
		if c.want != get {
			t.Errorf("unexpected response: want:%+v, get:%+v", c.want, get)
		}
	}
}
```

**Comment by wallacepf** [comment]
Tried this last example but seems like my get isn't returning any value, below you will find the entire code:

```
package cmd

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"

	"github.com/spf13/cobra"
)

// Wallet Object
type Wallet struct {
	Name string   `json:"name"`
	Tags []string `json:"tags"`
}

var wlt *Wallet

var tags []string
var tFlag string

func createWalletCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "wallet",
		Short: "Create a new wallet",
		Args:  cobra.ExactArgs(1),
		PreRun: func(cmd *cobra.Command, args []string) {
			wlt = new(Wallet)
			wlt.Name = args[0]

			if len(tags) == 0 {
				wlt.Tags = append(wlt.Tags, "default")
			} else {
				for _, v := range tags {

					wlt.Tags = append(wlt.Tags, v)
				}
			}

			if tFlag != "enabled" {

				fmt.Println("Wallet:", wlt.Name, "\nTags:", wlt.Tags, "\nIs this operation correct? y/N")

				var decision string

				fmt.Scan(&decision)
				if decision != "y" {
					log.Fatal("Wallet Insertion Cancelled", wlt.Name, tFlag)
				}
			} else {
				fmt.Println("OK")
			}

		},
		Run: func(cmd *cobra.Command, args []string) {

			jsonData, _ := json.Marshal(wlt)

			response, err := http.Post(appURL+"wallet", "application/json", bytes.NewBuffer(jsonData))
			if err != nil {
				// fmt.Printf("ok")
				if tFlag == "enabled" {
					fmt.Printf("OK")
				} else {
					log.Fatal(err)
				}

			} else {
				data, _ := ioutil.ReadAll(response.Body)
				fmt.Println(string(data))
			}

		},
	}
	cmd.Flags().StringVar(&tFlag, "testing", "", "for testing")
	cmd.Flags().MarkHidden("testing")
	cmd.Flags().StringSliceVarP(&tags, "tags", "t", []string{}, "Add tags to the wallet (For more than one, use csv)")

	return cmd
}

func init() {
	createCmd.AddCommand(createWalletCmd())
}
```

my test file

```
package cmd

import (
	"bytes"
	"testing"

	"github.com/mattn/go-shellwords"
)

func TestShow(t *testing.T) {
	cases := []struct {
		command string
		want    string
		errWant string
	}{
		{command: "wallet MyWallet --testing enabled", want: "OK", errWant: ""},
	}

	for _, c := range cases {
		buf := new(bytes.Buffer)
		cmd := createWalletCmd()
		cmd.SetOutput(buf)
		cmdArgs, err := shellwords.Parse(c.command)
		if err != nil {
			t.Fatalf("args parse error: %+v\n", err)
		}
		cmd.SetArgs(cmdArgs[1:])
		if err := cmd.Execute(); err != nil {
			if c.errWant != err.Error() {
				t.Errorf("unexpected error response: errWant: %+v, get:%+v", c.errWant, err.Error())
			}
		}

		get := buf.String()
		if c.want != get {
			t.Errorf("unexpected response: want: %+v, get: %+v", c.want, get)
		}
	}
}

```

the output after running the test

```
Running tool: /usr/local/bin/go test -timeout 30s -run ^TestShow$ wallet-cli/cmd

OK
OK--- FAIL: TestShow (0.00s)
    /Users/wallacepf/Dev/wallet project/go_wallet_cli/cmd/create_wallet_test.go:36: unexpected response: want: OK, get: 
FAIL
FAIL	wallet-cli/cmd	0.518s
FAIL
```

These "OK" were printed from my cobra command and that's why it gets really odd. Even though my script is returning twice the OK (One in the prerun and the other one in the Run) it's not taken in consideration. Both OKs were added just to validate the test ok? JFYI

**Comment by KEINOS** [comment]
@wallacepf 

>  seems like my get isn't returning any value

OMG. The same thing happened to me. The `buf` isn't buffering anything.

But the [original sample](https://github.com/tkit/go-cmd-example) did work. And I found that `cobra` version differs from ours. (See below)

<details><summary>Details running the sample</summary><div>

```shellsession
$ git clone https://github.com/tkit/go-cmd-example.git
Cloning into 'go-cmd-example'...
remote: Enumerating objects: 38, done.
remote: Total 38 (delta 0), reused 0 (delta 0), pack-reused 38
Unpacking objects: 100% (38/38), done.

$ cd go-cmd-example
$ go mod init sample
go: creating new go.mod: module sample
go: copying requirements from Gopkg.lock

$ go install
....

$ go test ./...
?       sample  [no test files]
ok      sample/cmd      (cached)

$ cat go.mod | grep cobra
        github.com/spf13/cobra v0.0.1

$ cat go.mod
module sample

go 1.15

require (
        github.com/fsnotify/fsnotify v1.4.2
        github.com/go-playground/locales v0.11.2
        github.com/go-playground/universal-translator v0.16.0
        github.com/hashicorp/hcl v0.0.0-20171017181929-23c074d0eceb
        github.com/inconshreveable/mousetrap v1.0.0
        github.com/magiconair/properties v1.7.3
        github.com/mattn/go-shellwords v1.0.3
        github.com/mitchellh/go-homedir v0.0.0-20161203194507-b8bc1bf76747
        github.com/mitchellh/mapstructure v0.0.0-20171017171808-06020f85339e
        github.com/pelletier/go-toml v1.0.1
        github.com/spf13/afero v1.0.0
        github.com/spf13/cast v1.1.0
        github.com/spf13/cobra v0.0.1
        github.com/spf13/jwalterweatherman v0.0.0-20170901151539-12bd96e66386
        github.com/spf13/pflag v1.0.0
        github.com/spf13/viper v1.0.0
        github.com/tkit/go-cmd-example v0.0.0-20171126084729-c013a3f0a738
        golang.org/x/sys v0.0.0-20171124171047-4ff8c001ce4c
        golang.org/x/text v0.1.1-0.20171102192421-88f656faf3f3
        gopkg.in/go-playground/validator.v9 v9.9.0
        gopkg.in/yaml.v2 v2.0.0-20171116090243-287cf08546ab
)
```

</div></details>

The whole problem is:
- The sample and ours both use the **currently deprecated function [`SetOutput()`](https://github.com/spf13/cobra/blob/master/command.go#L235-L236)** ...

I switched from `SetOutput()` to `SetOut()` and did work. Hope this helps!

**Comment by KEINOS** [comment]
After some coding, I started to feel `capturer.CaptureStdout()` is a lot easier to test and read, and short as well.

**Comment by wallacepf** [comment]
@KEINOS so first of all I have to say thank you very much to get engaged with me in this adventure hehehe. I could fix the code and properly get it tested and validated. What I did to get it properly working was:

```
package cmd

import (
	"bytes"
	"testing"

	"github.com/kami-zh/go-capturer"
	"github.com/stretchr/testify/assert"
)

func TestShow(t *testing.T) {
	var contains string
	var cmdTmp = createWalletCmd()
	var argsTmp = []string{"wallet teste", "--testing", "enabled"}
	buf := new(bytes.Buffer)
	result := capturer.CaptureStdout(func() {
		cmdTmp.SetOut(buf)
		cmdTmp.SetArgs(argsTmp)
		cmdTmp.Execute()
	})
	contains = "OK"

	assert.Equal(t, result, contains, "Not OK")

}

```

and it simply worked as expected! Once again, thank you very much and I believe you can close the case.

**Comment by KEINOS** [comment]
Nice! and clean! Glad that you solved it!

And because of this, I also got a little deeper into `cobra` and `go`, thank you too!

I found that searching the answer together like this one let us learn.

So I opened the GitHub [Discussions](https://github.com/KEINOS/Hello-Cobra/discussions) tab above. **Feel free to share questions** again there! 

- For some reason, I couldn't move this issue to the `Discussions`, so I'll leave this question here in the issue.

Now closing! :tada:
